synapse/rest/client/v1/room.py

```diff
@@ -38,6 +38,7 @@
     parse_integer,
```

Comment:
Historical state thread

Reply:
I need to look into inserting historical state (`PUT /state`) so the user meta info is available when I insert the historical message. But doesn't need to necessarily be in the current room state. Since we have to go from most recent to oldest using the `backfilled` option, I think I have to insert a membership event after the message was created so I wonder how that will work out.

But in general need to look into `backfilled` state on how that works (may be different).

Reply:
I think maybe the easiest is to include the state at the start of a chunk in any new "bulk send" API.

Reply:
Playing around with trying to insert a message from a user not already in the room has proven the point. The bulk send API is necessary if we're using the `backfilled = true` Synapse implementation which requires reverse-chronological history insertion.

If we try  to use the single send endpoint, we can't insert a `m.room.member` before the actual histortical message. If we insert a `m.room.member` back in time, a subsequent message has to come before it in time so it wont' resolve that state that came after.

---

For reference, if you register and join the user to the room as normal (so the current room state has the historical user), then try to backfill a historical message, it will give you a `SynapseError: 403 - User @maria:hs1 not in room !ZIIflEwiKrDFeMEMKV:hs1`

Reply:
I've summarized a bunch of this knowledge on the MSC where there was already some discussion on this problem, https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r595458729. Hopefully not too in the weeds and confusing. The path forward isn't as clear to me and either direction could work.

Reply:
@erikjohnston Putting in your review queue to look at https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r595458729. You have much stronger and informed opinions on the structure of Synapse

Reply:
@erikjohnston Curious about if state will resolve in these proposed ideas: https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r597063114

Reply:
@erikjohnston How to create floating outlier events? https://github.com/matrix-org/matrix-doc/pull/2716#discussion_r600238863